File: tributary-client/test/hash-compatibility.test.ts
import { describe, it, expect } from 'vitest';
import { computeHash } from '../src/hashUtils';

// Test to verify hash functions between client and server produce identical results

describe('Hash Function Compatibility Test', () => {
  it('should produce consistent hash results using the same algorithm', async () => {
    // Test data
    const testData = new TextEncoder().encode('test data for hashing');
    const priorHash = 'abc123def456';
    
    // Compute body hash using our implementation
    const bodyHash = await computeHash(testData);
    
    // Compute the chained hash (SHA256(priorHash + bodyHash)) - same as server would do after fix
    const concatenated = `${priorHash}${bodyHash}`;
    const chainedHash = await computeHash(new TextEncoder().encode(concatenated));
    
    // Verify the hash is computed consistently
    // For "test data for hashing", SHA256 should be consistent
    const expectedBodyHash = 'f7eb7961d8a233e6256d3a6257548bbb9293c3a08fb3574c88c7d6b429dbb9f5';
    const expectedChainedHash = 'e8910954652f2957dd5b6f34d88c78ff7f086546e2b94aef687290d409519a67';
    
    expect(bodyHash).toBe(expectedBodyHash);
    expect(chainedHash).toBe(expectedChainedHash);
  });
  
  it('should validate the hash computation process is deterministic', async () => {
    // Simulate the exact process used by both client and server after fix
    
    // Step 1: Prepare test data
    const testData = new TextEncoder().encode('Hello, Tributary!');
    const priorHash = 'a1b2c3d4e5f';
    
    // Step 2: Compute body hash (SHA256 of data)
    const bodyHash = await computeHash(testData);
    
    // Step 3: Compute chained hash (SHA256(priorHash + bodyHash))
    const concatenated = `${priorHash}${bodyHash}`;
    const chainedHash = await computeHash(new TextEncoder().encode(concatenated));
    
    // Both should use the same algorithm, same encoding (hex), same chaining
    
    // For "Hello, Tributary!", the SHA256 should be:
    const expectedBodyHash = '692f392e53f691be69dd1e502ef474a9103a19a48ef7a0a9115ee83d3a4bcb57';
    const expectedChainedHash = 'c6a7678f0c10c4ef797589575b0c8ffc108ee965cf1c06fef71cec3edc867b91';
    
    expect(bodyHash).toBe(expectedBodyHash);
    expect(chainedHash).toBe(expectedChainedHash);
  });
  
  it('should match the exact same process used by tributary-server after fix', async () => {
    // This test verifies that our implementation matches exactly what tributary-server does after the fix
    
    // The server now uses sha2 crate to:
    // 1. Compute SHA256 of body data
    // 2. Concatenate prior_hash + body_hash
    // 3. Compute SHA256 of the concatenated result
    
    // Test with empty data (edge case)
    const emptyData = new TextEncoder().encode('');
    const emptyBodyHash = await computeHash(emptyData);
    const expectedEmptyBodyHash = 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855';
    expect(emptyBodyHash).toBe(expectedEmptyBodyHash);
    
    // Test computing the chained hash for empty data
    const emptyPriorHash = '';
    const emptyConcatenated = `${emptyPriorHash}${emptyBodyHash}`;
    const emptyChainedHash = await computeHash(new TextEncoder().encode(emptyConcatenated));
    // This would be SHA256("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855")
    // = SHA256(empty string hash) = "057dba3d05308af630d801001bc94eeecc6a65142e9d1e88ecc713b3a908f1ab"
    
    // Test with typical blob data
    const blobData = new TextEncoder().encode('Sample blob data for tributary');
    const blobBodyHash = await computeHash(blobData);
    
    // Verify the process: SHA256(prior_hash + body_hash)
    const priorHash = 'prev-hash-123';
    const concatenated = `${priorHash}${blobBodyHash}`;
    const chainedHash = await computeHash(new TextEncoder().encode(concatenated));
    
    // This is the exact same process the Rust server uses in compute_hash + hash chaining
    expect(typeof chainedHash).toBe('string');
    expect(chainedHash).toHaveLength(64); // SHA256 produces 64-character hex string
  });
});

File: tributary-server/tests/hash_compatibility_test.rs
// This file demonstrates the exact hash process used by tributary-server
// and verifies that tributary-client produces identical results

use sha2::{Digest, Sha256};

/// Compute SHA256 hash of data - matches tributary-client implementation
/// This function demonstrates the exact same process used in compute_hash in tributary-server/src/crypto.rs
pub fn compute_hash(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let result = hasher.finalize();
    hex::encode(result)
}

/// Compute chain hash (SHA256(prior_hash + body_hash)) - matches updated tributary-client implementation
/// This function demonstrates the exact same process used in tributary-server/src/api.rs after the fix
pub fn compute_chain_hash(prior_hash: &str, body_data: &[u8]) -> String {
    // Compute body hash first
    let body_hash = compute_hash(body_data);

    // Concatenate prior_hash + body_hash, then hash the result
    let concatenated = format!("{}{}", prior_hash, body_hash);
    compute_hash(concatenated.as_bytes())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hash_compatibility_with_client() {
        // Test the exact same data that the client test uses
        let test_data = "test data for hashing".as_bytes();
        let prior_hash = "abc123def456";

        // Compute using server implementation
        let body_hash = compute_hash(test_data);
        let chain_hash = compute_chain_hash(prior_hash, test_data);

        // These should match exactly what the client produces:
        // body_hash should be "f7eb7961d8a233e6256d3a6257548bbb9293c3a08fb3574c88c7d6b429dbb9f5"
        // chain_hash should be SHA256(prior_hash + body_hash) = SHA256("abc123def456f7eb7961d8a233e6256d3a6257548bbb9293c3a08fb3574c88c7d6b429dbb9f5")
        // = "e8910954652f2957dd5b6f34d88c78ff7f086546e2b94aef687290d409519a67"

        assert_eq!(
            body_hash,
            "f7eb7961d8a233e6256d3a6257548bbb9293c3a08fb3574c88c7d6b429dbb9f5"
        );
        assert_eq!(
            chain_hash,
            "e8910954652f2957dd5b6f34d88c78ff7f086546e2b94aef687290d409519a67"
        );
    }

    #[test]
    fn test_hello_tributary() {
        let test_data = "Hello, Tributary!".as_bytes();
        let prior_hash = "a1b2c3d4e5f";

        let body_hash = compute_hash(test_data);
        let chain_hash = compute_chain_hash(prior_hash, test_data);

        // These should match exactly what the client produces:
        // body_hash should be "692f392e53f691be69dd1e502ef474a9103a19a48ef7a0a9115ee83d3a4bcb57"
        // chain_hash should be SHA256(prior_hash + body_hash) = SHA256("a1b2c3d4e5f692f392e53f691be69dd1e502ef474a9103a19a48ef7a0a9115ee83d3a4bcb57")
        // = "c6a7678f0c10c4ef797589575b0c8ffc108ee965cf1c06fef71cec3edc867b91"

        assert_eq!(
            body_hash,
            "692f392e53f691be69dd1e502ef474a9103a19a48ef7a0a9115ee83d3a4bcb57"
        );
        assert_eq!(
            chain_hash,
            "c6a7678f0c10c4ef797589575b0c8ffc108ee965cf1c06fef71cec3edc867b91"
        );
    }

    #[test]
    fn test_chain_hash_without_prior() {
        // Test chain hash computation when there is no prior hash (first entry)
        let test_data = "First entry".as_bytes();
        let prior_hash = ""; // Empty prior hash for first entry

        let body_hash = compute_hash(test_data);
        let chain_hash = compute_chain_hash(prior_hash, test_data);

        // For first entry with empty prior hash:
        // body_hash should be "7749435cc893289da9df793cdb29ba90e082e57e6a60f4019b1d22f57bc3bf40"
        // chain_hash should be SHA256("" + body_hash) = SHA256("7749435cc893289da9df793cdb29ba90e082e57e6a60f4019b1d22f57bc3bf40")
        // = "f962b7ec0d0375d2ee951857e1209bd1c6f70b26626face9c948569718503641"

        assert_eq!(
            body_hash,
            "7749435cc893289da9df793cdb29ba90e082e57e6a60f4019b1d22f57bc3bf40"
        );
        assert_eq!(
            chain_hash,
            "f962b7ec0d0375d2ee951857e1209bd1c6f70b26626face9c948569718503641"
        );
    }

    #[test]
    fn test_chain_hash_with_real_prior() {
        // Test chain hash computation with a real prior hash (simulating actual chaining)
        let test_data = "Second entry".as_bytes();
        let prior_hash = "f962b7ec0d0375d2ee951857e1209bd1c6f70b26626face9c948569718503641"; // hash from previous test

        let body_hash = compute_hash(test_data);
        let chain_hash = compute_chain_hash(prior_hash, test_data);

        // This tests actual chaining with real hashes, not faked ones
        // body_hash should be "f5f5fd73d02a1535460e64279b2aa672309c80b129f28efe2ab523f32d1e91be"
        // chain_hash should be SHA256(prior_hash + body_hash) = SHA256("f962b7ec0d0375d2ee951857e1209bd1c6f70b26626face9c948569718503641f5f5fd73d02a1535460e64279b2aa672309c80b129f28efe2ab523f32d1e91be")
        // = "b17d1dcc5867c9dbaf798b1d7cd91168a362702d4acfd0f8f997c7390aaf65f9"

        assert_eq!(
            body_hash,
            "f5f5fd73d02a1535460e64279b2aa672309c80b129f28efe2ab523f32d1e91be"
        );
        assert_eq!(
            chain_hash,
            "b17d1dcc5867c9dbaf798b1d7cd91168a362702d4acfd0f8f997c7390aaf65f9"
        );
    }
}
